api/src/main/java/org/apache/iceberg/metrics/ScanReport.java

```diff
@@ -0,0 +1,265 @@
+      return this;
+    }
+
+    public Builder withFilter(Expression newFilter) {
+      this.filter = newFilter;
+      return this;
+    }
+
+    public Builder withProjection(Schema newProjection) {
+      this.projection = newProjection;
+      return this;
+    }
+
+    public Builder fromScanMetrics(ScanMetrics newScanMetrics) {
+      this.scanMetrics = newScanMetrics;
+      return this;
+    }
+
+    public ScanReport build() {
+      Preconditions.checkArgument(null != tableName, "Invalid table name: null");
+      Preconditions.checkArgument(null != filter, "Invalid expression filter: null");
+      Preconditions.checkArgument(null != projection, "Invalid schema projection: null");
+      Preconditions.checkArgument(null != scanMetrics, "Invalid scan metrics: null");
+      return new ScanReport(tableName, snapshotId, filter, projection, scanMetrics);
+    }
+  }
+
+  /** Carries all metrics for a particular scan */
+  public static class ScanMetrics {
+    public static final ScanMetrics NOOP = new ScanMetrics(MetricsContext.nullMetrics());
+    private final Counter<Integer> resultDataFiles;
+    private final Counter<Integer> totalDataManifests;
+    private final Counter<Integer> totalDeleteManifests;
+    private final Counter<Integer> scannedDataManifests;
+    private final Counter<Integer> scannedDeleteManifests;
+    private final Counter<Integer> totalDataFiles;
+    private final Counter<Integer> totalDeleteFiles;
+    private final Timer totalPlanningDuration;
+    private final Counter<Long> totalFileSizeInBytes;
+    private final Counter<Integer> indexedDeleteFiles;
+    private final Counter<Integer> resultDeleteFiles;
+    private final Counter<Integer> skippedDataManifests;
+    private final Counter<Integer> skippedDeleteManifests;
+    private final Counter<Integer> skippedDataFiles;
+    private final Counter<Integer> skippedDeleteFiles;
+
+    public ScanMetrics(MetricsContext metricsContext) {
+      Preconditions.checkArgument(null != metricsContext, "Invalid metrics context: null");
+      this.totalDataManifests =
+          metricsContext.counter("totalDataManifests", Integer.class, MetricsContext.Unit.COUNT);
+      this.totalDeleteManifests =
+          metricsContext.counter("totalDeleteManifests", Integer.class, MetricsContext.Unit.COUNT);
+      this.resultDataFiles =
+          metricsContext.counter("resultDataFiles", Integer.class, MetricsContext.Unit.COUNT);
+      this.scannedDataManifests =
+          metricsContext.counter("scannedDataManifests", Integer.class, MetricsContext.Unit.COUNT);
+      this.scannedDeleteManifests =
+          metricsContext.counter(
+              "scannedDeleteManifests", Integer.class, MetricsContext.Unit.COUNT);
+      this.totalDataFiles =
+          metricsContext.counter("totalDataFiles", Integer.class, MetricsContext.Unit.COUNT);
+      this.totalDeleteFiles =
+          metricsContext.counter("totalDeleteFiles", Integer.class, MetricsContext.Unit.COUNT);
```

Comment:
I think this needs to be implemented. Currently, the code counts deleted files rather than live files in delete manifests. You should be able to update this in `DeleteFileIndex` by passing `scanMetrics` to the `ManifestReader`. I think that will also fix `skippedDeleteFiles`.